v4/supervisor.go

```diff
@@ -0,0 +1,956 @@
+	namedService, present := s.services[id]
+	if present {
+		cancel := s.cancellations[id]
+		delete(s.services, id)
+		delete(s.cancellations, id)
+
+		s.servicesShuttingDown[id] = namedService
+		go func() {
+			successChan := make(chan struct{})
+			go func() {
+				cancel()
+				close(successChan)
+				if notificationChan != nil {
+					notificationChan <- struct{}{}
+				}
+			}()
+
+			select {
+			case <-successChan:
+				// Life is good!
+			case <-s.getAfterChan(s.spec.Timeout):
+				s.spec.LogBadStop(
+					s,
+					namedService.Service,
+					namedService.name,
+				)
+			}
+			s.notifyServiceDone <- id
+		}()
+	} else {
+		if notificationChan != nil {
+			notificationChan <- struct{}{}
+		}
+	}
+}
+
+func (s *Supervisor) stopSupervisor() UnstoppedServiceReport {
+	notifyDone := make(chan serviceID, len(s.services))
+
+	for id := range s.services {
+		namedService, present := s.services[id]
+		if present {
+			cancel := s.cancellations[id]
+			delete(s.services, id)
+			delete(s.cancellations, id)
+			s.servicesShuttingDown[id] = namedService
+			go func(sID serviceID) {
+				cancel()
+				notifyDone <- sID
+			}(id)
+		}
+	}
+
+	timeout := s.getAfterChan(s.spec.Timeout)
+
+SHUTTING_DOWN_SERVICES:
+	for len(s.servicesShuttingDown) > 0 {
+		select {
+		case id := <-notifyDone:
+			delete(s.servicesShuttingDown, id)
+		case serviceID := <-s.notifyServiceDone:
+			delete(s.servicesShuttingDown, serviceID)
+		case <-timeout:
+			for _, namedService := range s.servicesShuttingDown {
+				s.spec.LogBadStop(
+					s,
+					namedService.Service,
+					namedService.name,
+				)
+			}
+
+			// failed remove statements will log the errors.
+			break SHUTTING_DOWN_SERVICES
+		}
+	}
+
+	// If nothing else has cancelled our context, we should now.
+	s.myCancel()
+
+	// Indicate that we're done shutting down
+	defer close(s.liveness)
+
+	if len(s.servicesShuttingDown) == 0 {
+		return nil
+	} else {
+		report := UnstoppedServiceReport{}
+		for serviceID, serviceWithName := range s.servicesShuttingDown {
+			report = append(report, UnstoppedService{
+				SupervisorPath: []*Supervisor{s},
+				Service:        serviceWithName.Service,
+				Name:           serviceWithName.name,
+				ServiceToken:   ServiceToken{uint64(s.id)<<32 | uint64(serviceID)},
+			})
+		}
+		s.m.Lock()
+		s.unstoppedServiceReport = report
+		s.m.Unlock()
+		return report
+	}
+}
+
+// String implements the fmt.Stringer interface.
+func (s *Supervisor) String() string {
+	return s.Name
+}
+
+// sendControl abstracts checking for the supervisor to still be running
+// when we send a message. This prevents blocking when sending to a
+// cancelled supervisor.
+func (s *Supervisor) sendControl(sm supervisorMessage) error {
+	var doneChan <-chan struct{}
+	s.ctxMutex.Lock()
+	if s.ctx == nil {
+		s.ctxMutex.Unlock()
+		return ErrSupervisorNotStarted
+	}
+	doneChan = s.ctx.Done()
+	s.ctxMutex.Unlock()
+
+	select {
+	case s.control <- sm:
+		return nil
+	case <-doneChan:
+		return ErrTimeout
```

Comment:
Using `ErrTimeout` to represent that supervisor was stopped/context was cancelled feels like a source of confusion. I'd create a dedicated error like `ErrSupervisorNotStarted`.